### DISCUSSION TITLE
Avalonia.Browser retrieve app data folder path

### DISCUSSION THREAD

**Discussion by ClaudiuHBann** [AvaloniaUI/Avalonia#14119] [Q&A] [answered]
Is there a way of getting the app data path in the browser?
For desktop applications I use `Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData)` and it works, but for the browser it returns an empty string.
And if we find a way of getting the path, are we allowed to read/write from/to it in browser?

**Comment by thevortexcloud** [comment]
You specifically want `%appdata%`? If so that's not going to work. WASM is a generic target that is limited to the sandbox provided by the environment (typically a web browser). Direct access to the local file system is heavily restricted for several reasons. Use a desktop app if you want to do this.

**Reply by ClaudiuHBann** [reply]
i have a desktop app too but I would very much like a browser version of my application, is there any way of caching some data in the browser, any tutorial or smth, if this approach doesn't work?

**Reply by maxkatz6** [reply]
HttpClient requests might be cached depending on the content type and cache headers. But you won’t have much control over that, as Browser is responsible for that. 

IndexedDB is another option available in the browser to caching big data, where you have a bit more control. 

In general, this whole discussion has nothing to do with Avalonia, and you can do a research with general browser documentations. Like this https://developer.mozilla.org/en-US/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria

**Reply by thevortexcloud** [reply]
It depends on what you are trying to cache. Modern browsers provide several ways of caching and persisting data.

https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Client-side_storage

Alternatively you can do the classical way of doing this and set up some sort of server/API your app can use which will handle all the caching and persistence for your client.

**Reply by ClaudiuHBann** [reply]
I want to cache some settings for my app like `Show something` or cache an account details name, password etc...
What's the best options for this? And it there an example of that option you know about?
BTW, much thanks for helping me out guys! <3

**Reply by thevortexcloud** [reply, marked as answer]
Do not cache passwords. It's a bad idea. If you are doing any kind of sending a password over the network, the server should instead issue some sort of token when valid credentials are presented, which can then be cached. Anyway if you just want a simple key value storage then yes, you can use the Storage API.

https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API

Frankly though it sounds more like you want a proper server side database. One of the major rules of web development is "**NEVER** trust the client".

**Reply by ClaudiuHBann** [reply]
Thank you for everything, much love!

**Reply by thevortexcloud** [reply]
You are welcome!
